### ISSUE TITLE
Auto-generate rest client

### ISSUE THREAD

**Issue by callumforrester** [DiamondLightSource/blueapi#364]
We should auto-generate the rest client code rather than handcraft it. Replacing `blueapi.cli.rest.BlueapiRestClient` with something generated with swagger-codegen or similar

**Comment by stan-dot** [comment]
todo - make a test repository demonstrating proof of concept with fastapi
https://fastapi.tiangolo.com/advanced/generate-clients/

**Comment by stan-dot** [comment]
@callumforrester need more info on the use case for this client. Will it be called from a Python GUI app? Right now it's only the CLI client. 

most generators focus on the Typescript use case. The documentation for the Python one isn't that great https://openapi-generator.tech/docs/generators/python/

**Comment by stan-dot** [comment]
ah I see there's a huge benefit from autocompletion - but I'm not sure if I got it running on my device

**Comment by stan-dot** [comment]
what is the target use case for building this CLI? Testing the API? running it? I've looked into autogeneration and there are libraries for that, but then we'd need to write custom CLI to it. I haven't noticed a library that would produce the CLI at the same time. We would need then to manually update the cli code on every regeneration. there is a plethora of tools for rapid api testing or calling remote servers. They are much more powerful than custom CLI, with more support. 

If we go with recommending a GUI option like Postman, users could even have an account and keep their API call data across devices. If not Postman we can consider some alternatives:
- [httpie ](https://github.com/httpie/cli)
- [jmeter from apache](https://github.com/apache/jmeter)
- [testsigma](https://testsigma.com/automated-api-testing) - more geared towards automation
- [insomnia](https://github.com/Kong/insomnia) - with also support for git storage, websockets and grpc
- [soapUI](https://github.com/SmartBear/soapui) 
- [redocly](https://github.com/Redocly/redocly-cli) - not necessarily a rest client, but a good tool to use with openapi
- [postman windows client](https://github.com/portapps/postman-portable) - this could be huge for many users

In conclusion, using a third party instead of in house CLI would be beneficial.

In regards to the client for a GUI python / web app that can be done quite easily. I will add a github action workflow soon

**Comment by callumforrester** [comment]
For now I think the only use case is auto-generating a version of this class: https://github.com/DiamondLightSource/blueapi/blob/7bbc94e0d61da2a4ce4de6a1285c4cc0e4ba67f2/src/blueapi/cli/rest.py#L26

Future classes that fulfil the use cases you suggest can build on top of it.

**Comment by stan-dot** [comment]
I mean use case in the broad sense of having the Python Rest client in the first place

**Comment by callumforrester** [comment]
Yes, the use case is mainly testing/debugging. The reason for writing this and not just using rest tools is that 
the CLI doesn't just make rest calls though, it combines several rest calls together for the user's convenience and also listens to the message bus via a query-then-subscribe model.

**Comment by stan-dot** [comment]
if the client is generated, we could publish it as a pypi package and have cli as a separate project. admittedly if Kubernetes is the inspiration, they have a build setup with a monorepo.

depends on what is the target use case I guess

https://github.com/kubernetes/kubernetes/blob/master/build/root/Makefile

**Comment by stan-dot** [comment]
> For most cases, especially in a microservices architecture or when sharing clients among multiple teams, the registry approach aligns better with continuous integration and delivery practices. It optimizes reuse and version management, although it requires a bit more setup initially.

chatgpt has an opinion

**Comment by stan-dot** [comment]
note: java has a build repo in gitlab maven using this dynamically, with version set manually on release. not scalable

**Comment by stan-dot** [comment]
note: this setup has a tighter loop for DX rather that leaving this to CI. can be explored next https://chat.openai.com/share/eb5d3fb7-5024-4146-a50a-4cda7c37c70f

**Comment by stan-dot** [comment]
working on this now

**Comment by stan-dot** [comment]
the generator produces not one file, but a whole project:
`
![image](https://github.com/DiamondLightSource/blueapi/assets/56644812/e8910422-22aa-4522-8340-4c2514539b5d)
`

**Comment by stan-dot** [comment]
note from discussion - to [reuse the structure from the tiled repo.](https://github.com/bluesky/tiled/tree/main/tiled/client)

**Comment by stan-dot** [comment]
based on the #449 pydantic 2 migration we'll be able to use the [generic python generator](https://openapi-generator.tech/docs/generators/python/) instead of the odd pydantic-v1 version. It makes no sense to migrate at the moment to v1 if after #449 we'd need to migrate again. delaying this right now
